src/main/scala/akka/stream/contrib/Retry.scala

```diff
@@ -0,0 +1,245 @@
+   * indevidually. Since it's very easy to inflate elements in retry cycle, there's also a
+   * limit parameter, that will limit the amount of generated elements by the `retryWith` function,
+   * and will fail the stage if that limit is exceeded.
+   * Passing `Some(Nil)` is valid, and will result in filtering out the failure quietly, without
+   * emitting a failed `Try` element.
+   *
+   * IMPORTANT CAVEAT:
+   * The given flow must not change the number of elements passing through it (i.e. it should output
+   * exactly one element for every recieved element). Ignoring this, will have an unpredicted result.
+   *
+   * @param limit since every retry can generate more elements,
+   *        the inner queue can get too big. if the limit is reached,
+   *        the stage will fail.
+   * @param flow the flow to retry
+   * @param retryWith if output was failure, we can optionaly recover from it,
+   *        and retry with a sequence of input & new state pairs we get from this function.
+   * @tparam I input elements type
+   * @tparam O output elements type
+   * @tparam S state to create a new (input,state): (I,S) to retry with
+   * @tparam M materialized value type
+   * @return `Graph[FlowShape[(I,S),Try[O]],M]` instance
+   */
+  def concat[I, O, S, M](limit: Int, flow: Graph[FlowShape[(I, S), (Try[O], S)], M])(retryWith: S => Option[immutable.Iterable[(I, S)]]): Graph[FlowShape[(I, S), Try[O]], M] = {
+    GraphDSL.create(flow) { implicit b => origFlow =>
+      import GraphDSL.Implicits._
+
+      val elemQueue = b.add(new QueueElementsAndInterceptCompletion[(I, S)](limit))
+      val partition = b.add(PartitionWith[(Try[O], S), Try[O], immutable.Iterable[(I, S)]] {
+        case (success: Success[O], _) => Left(success)
+        case (failedTry, s)           => retryWith(s).fold[Either[Try[O], immutable.Iterable[(I, S)]]](Left(failedTry))(Right.apply)
+      })
+      val coElemQueue = b.add(new InterceptElemntsLeavingCycle[Try[O]])
+      // format: OFF
+      elemQueue.out ~> origFlow ~> partition.in
+                                   partition.out0 ~> coElemQueue.in
+      elemQueue.in1       <~       partition.out1
+      elemQueue.in2                  <~              coElemQueue.out1
+      // format: ON
+      FlowShape(elemQueue.in0, coElemQueue.out0)
+    }
+  }
+
+  private[akka] class CountElementsAndInterceptCompletion[T] extends GraphStage[FanInShape2[T, Int, T]] {
+    override val shape = new FanInShape2[T, Int, T]("countElementsAndInterceptCompletion")
+    override def createLogic(attributes: Attributes) = new GraphStageLogic(shape) {
+
+      import shape._
+      var elementsInCycleCount = 0
+
+      override def preStart() = pull(in1)
+
+      setHandler(in0, new InHandler {
+        override def onPush() = {
+          push(out, grab(in0))
+          elementsInCycleCount += 1
+        }
+
+        override def onUpstreamFinish() = {
```

Comment:
I don't think so. in case there are no elements in the cycle, the stage should complete regularly.